utils/console_menu.py

```diff
@@ -0,0 +1,209 @@
+#!/usr/bin/env micropython
+from time import sleep
+from sys import stderr
+from os import listdir
+from ev3dev2.button import Button
+from ev3dev2.console import Console
+from ev3dev2.led import Leds
+from ev3dev2.sensor import list_sensors, INPUT_1, INPUT_2, INPUT_3, INPUT_4
+
+
+"""
+Used to create a console menu for switching between programs quickly
+without having to return to Brickman to find and launch a program.
+Demonstrates the EV3DEV2 Console(), Led(), and Button() classes.
+
+"""
+
+
+def get_positions(console):
+    """
+    Compute a dictionary keyed by button names with horizontal alignment,
+    and column/row location to show each choice on the EV3 LCD console.
+
+    Parameter:
+
+    - `console (Console): an instance of the EV3 Console() class
+
+    returns a dictionary keyed by button names with column/row location
+    """
+
+    midrow = 1 + console.rows // 2
+    midcol = 1 + console.columns // 2
+    # horiz_alignment, col, row
+    return {
+        "up": ("C", midcol, 1),
+        "right": ("R", console.columns, midrow),
+        "down": ("C", midcol, console.rows),
+        "left": ("L", 1, midrow),
+        "enter": ("C", midcol, midrow)
+    }
+
+
+def wait_for_button_press(button):
+    """
+    Wait for a button to be pressed and released.
+
+    Parameter:
+
+    - `button (Button): an instance of the EV3 Button() class
+
+    return the Button name that was pressed.
+    """
+    pressed = None
+    while True:
+        allpressed = button.buttons_pressed
+        if bool(allpressed):
+            pressed = allpressed[0]  # just get the first one
+            while not button.wait_for_released(pressed):
+                pass
+            break
+    return pressed
+
+
+def menu(choices, before_run_function=None, after_run_function=None):
+    """
+    Console Menu that accepts choices and corresponding functions to call.
+    The user must press the same button twice: once to see their choice highlited,
+    a second time to confirm and run the function. The EV3 LEDs show each state change:
+    Green = Ready for button, Amber = Ready for second button, Red = Running
+
+    Parameters:
+
+    - `choices` a dictionary of tuples "button-name": ("mission-name", function-to-call)
```

Comment:
I don't think this has to be specific to missions; one could use this for other menus, right?

Reply:
Sure... its a nifty routine to tie-in functions to the EV3 buttons... and uses the console and LEDs to provide feedback when selecting and confirming a choice. For _FIRST_ LEGO League, we refer to the scoring runs as "missions"... and they can chain them together to accomplish multiple models in the same "sortie".

Reply:
Yes correct... I was thinking of a new util to be a better replacement to the clunky brickman show sensors tool... use the menu instead to show sensor values, operate the motors, etc. There should be a cool way to nest menus within a top-level menu to create a menu tree for quick navigation--may need a "single-press" vs. "double-press" option to make it even simpler.